File: ios/LLMHub/Sources/LLMHub/MCPClient.swift
import Foundation
import Security

struct MCPTool: Identifiable, Equatable {
    let name: String
    let description: String
    let inputSchema: [String: Any]
    var id: String { name }
    var exposedName: String {
        "mcp_" + name.map { $0.isLetter || $0.isNumber || $0 == "_" ? String($0) : "_" }.joined()
    }

    static func == (lhs: MCPTool, rhs: MCPTool) -> Bool { lhs.name == rhs.name }

    var promptLine: String {
        let properties = inputSchema["properties"] as? [String: Any] ?? [:]
        let required = Set(inputSchema["required"] as? [String] ?? [])
        let arguments = properties.keys.sorted { lhs, rhs in
            if required.contains(lhs) != required.contains(rhs) { return required.contains(lhs) }
            return lhs < rhs
        }.map { key in
            let schema = properties[key] as? [String: Any]
            return "\"\(key)\": \(Self.argumentExample(for: schema, required: required.contains(key)))"
        }.joined(separator: ", ")
        return "- \(exposedName)({\(arguments)}): \((description.isEmpty ? name : description).prefix(300))"
    }

    private static func argumentExample(for schema: [String: Any]?, required: Bool) -> String {
        let suffix = required ? " REQUIRED" : ""
        let type = schema?["type"] as? String ?? "value"
        switch type {
        case "string":
            return "\"<string\(suffix)>\""
        case "number", "integer":
            return "\"<number\(suffix)>\""
        case "boolean":
            return required ? "true" : "false"
        case "array":
            let itemType = (schema?["items"] as? [String: Any])?["type"] as? String ?? "value"
            return itemType == "string" ? "[\"<string>\"]" : "[\"<\(itemType)>\"]"
        case "object":
            return "{}"
        default:
            return "\"<\(type)\(suffix)>\""
        }
    }
}

struct MCPSettings: Equatable {
    var enabled: Bool
    var url: String
    var token: String
}

enum MCPClientError: LocalizedError {
    case message(String)
    var errorDescription: String? {
        if case .message(let value) = self { return value }
        return nil
    }
}

@MainActor
final class MCPClient {
    static let shared = MCPClient()
    private let currentProtocol = "2026-07-28"
    private let legacyProtocol = "2025-06-18"
    private var protocolVersion = "2026-07-28"
    private var sessionID: String?
    private var requestID = 1
    private(set) var tools: [MCPTool] = []

    private var appVersion: String {
        Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0"
    }

    private var clientInfo: [String: String] {
        ["name": "LLM Hub iOS", "version": appVersion]
    }

    func loadSettings() -> MCPSettings {
        MCPSettings(
            enabled: UserDefaults.standard.bool(forKey: "agent_mcp_enabled"),
            url: UserDefaults.standard.string(forKey: "agent_mcp_url") ?? "",
            token: MCPKeychain.read()
        )
    }

    func saveSettings(_ settings: MCPSettings) {
        UserDefaults.standard.set(settings.enabled, forKey: "agent_mcp_enabled")
        UserDefaults.standard.set(settings.url.trimmingCharacters(in: .whitespacesAndNewlines), forKey: "agent_mcp_url")
        MCPKeychain.write(settings.token.trimmingCharacters(in: .whitespacesAndNewlines))
    }

    func disconnect() {
        sessionID = nil
        protocolVersion = currentProtocol
        tools = []
    }

    func connectAndDiscover(settings: MCPSettings) async throws -> [MCPTool] {
        guard settings.enabled else { throw MCPClientError.message("MCP is disabled") }
        try validateURL(settings.url)
        disconnect()
        let result: [String: Any]
        do {
            result = try await rpc(settings: settings, method: "tools/list", params: [:], version: currentProtocol)
        } catch {
            protocolVersion = legacyProtocol
            let initialized = try await rpc(
                settings: settings,
                method: "initialize",
                params: [
                    "protocolVersion": legacyProtocol,
                    "capabilities": [:],
                    "clientInfo": clientInfo
                ],
                version: legacyProtocol,
                includeProtocolHeader: false
            )
            protocolVersion = initialized["protocolVersion"] as? String ?? legacyProtocol
            try await notify(settings: settings, method: "notifications/initialized")
            result = try await rpc(settings: settings, method: "tools/list", params: [:], version: protocolVersion)
        }
        let rawTools = result["tools"] as? [[String: Any]] ?? []
        tools = rawTools.prefix(64).compactMap { item in
            guard let name = item["name"] as? String, !name.isEmpty else { return nil }
            return MCPTool(name: name, description: item["description"] as? String ?? "", inputSchema: item["inputSchema"] as? [String: Any] ?? ["type": "object"])
        }
        return tools
    }

    func callTool(name: String, arguments: [String: Any], settings: MCPSettings) async throws -> String {
        let result = try await rpc(
            settings: settings,
            method: "tools/call",
            params: ["name": name, "arguments": arguments],
            version: protocolVersion,
            mcpName: name
        )
        var parts: [String] = []
        for item in result["content"] as? [[String: Any]] ?? [] {
            if item["type"] as? String == "text", let text = item["text"] as? String { parts.append(text) }
            else if let uri = item["uri"] as? String { parts.append(uri) }
        }
        if let structured = result["structuredContent"],
           let data = try? JSONSerialization.data(withJSONObject: structured, options: [.sortedKeys]),
           let text = String(data: data, encoding: .utf8) { parts.append(text) }
        if parts.isEmpty,
           let data = try? JSONSerialization.data(withJSONObject: result, options: [.sortedKeys]),
           let text = String(data: data, encoding: .utf8) {
            parts.append(text)
        }
        let output = String(parts.joined(separator: "\n").prefix(100_000))
        if result["isError"] as? Bool == true { throw MCPClientError.message(output.isEmpty ? "MCP tool failed" : output) }
        return output
    }

    private func validateURL(_ raw: String) throws {
        guard let url = URL(string: raw.trimmingCharacters(in: .whitespacesAndNewlines)), let scheme = url.scheme?.lowercased(), let host = url.host else {
            throw MCPClientError.message("Invalid MCP server URL")
        }
        guard scheme == "https" || (scheme == "http" && ["localhost", "127.0.0.1"].contains(host)) else {
            throw MCPClientError.message("MCP servers must use HTTPS")
        }
    }

    private func rpc(settings: MCPSettings, method: String, params: [String: Any], version: String, includeProtocolHeader: Bool = true, mcpName: String? = nil) async throws -> [String: Any] {
        let id = requestID
        requestID += 1
        var requestParams = params
        if version == currentProtocol {
            requestParams["_meta"] = ["io.modelcontextprotocol/clientInfo": clientInfo]
        }
        let response = try await execute(
            settings: settings,
            json: ["jsonrpc": "2.0", "id": id, "method": method, "params": requestParams],
            method: method,
            version: version,
            includeProtocolHeader: includeProtocolHeader,
            mcpName: mcpName
        )
        if let error = response["error"] as? [String: Any] {
            throw MCPClientError.message(error["message"] as? String ?? "MCP request failed")
        }
        guard let result = response["result"] as? [String: Any] else { throw MCPClientError.message("Invalid MCP response") }
        return result
    }

    private func notify(settings: MCPSettings, method: String) async throws {
        _ = try await execute(settings: settings, json: ["jsonrpc": "2.0", "method": method], method: method, version: protocolVersion, expectResponse: false)
    }

    private func execute(settings: MCPSettings, json: [String: Any], method: String, version: String, includeProtocolHeader: Bool = true, mcpName: String? = nil, expectResponse: Bool = true) async throws -> [String: Any] {
        guard let url = URL(string: settings.url.trimmingCharacters(in: .whitespacesAndNewlines)) else { throw MCPClientError.message("Invalid MCP server URL") }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 45
        request.httpBody = try JSONSerialization.data(withJSONObject: json)
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json, text/event-stream", forHTTPHeaderField: "Accept")
        request.setValue(method, forHTTPHeaderField: "Mcp-Method")
        if includeProtocolHeader { request.setValue(version, forHTTPHeaderField: "MCP-Protocol-Version") }
        if let mcpName { request.setValue(mcpName, forHTTPHeaderField: "Mcp-Name") }
        if let sessionID { request.setValue(sessionID, forHTTPHeaderField: "Mcp-Session-Id") }
        if !settings.token.isEmpty { request.setValue("Bearer \(settings.token)", forHTTPHeaderField: "Authorization") }
        let (data, rawResponse) = try await URLSession.shared.data(for: request)
        guard let response = rawResponse as? HTTPURLResponse else { throw MCPClientError.message("Invalid MCP response") }
        if let value = response.value(forHTTPHeaderField: "Mcp-Session-Id") { sessionID = value }
        guard (200...299).contains(response.statusCode) else {
            throw MCPClientError.message("HTTP \(response.statusCode): \(String(data: data.prefix(500), encoding: .utf8) ?? "")")
        }
        if !expectResponse || response.statusCode == 202 { return [:] }
        guard data.count <= 1_000_000 else { throw MCPClientError.message("MCP response is too large") }
        var payload = data
        if response.value(forHTTPHeaderField: "Content-Type")?.contains("text/event-stream") == true {
            let events = (String(data: data, encoding: .utf8) ?? "").split(separator: "\n").compactMap { line -> String? in
                guard line.hasPrefix("data:") else { return nil }
                return String(line.dropFirst(5)).trimmingCharacters(in: .whitespaces)
            }
            guard let event = events.last(where: { $0.contains("\"jsonrpc\"") }), let eventData = event.data(using: .utf8) else {
                throw MCPClientError.message("Empty MCP event stream")
            }
            payload = eventData
        }
        guard let object = try JSONSerialization.jsonObject(with: payload) as? [String: Any] else { throw MCPClientError.message("Invalid MCP JSON") }
        return object
    }
}

private enum MCPKeychain {
    private static let service = "com.llmhub.llmhub.mcp"
    private static let account = "bearer-token"

    static func read() -> String {
        let query: [String: Any] = [kSecClass as String: kSecClassGenericPassword, kSecAttrService as String: service, kSecAttrAccount as String: account, kSecReturnData as String: true]
        var result: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &result) == errSecSuccess, let data = result as? Data else { return "" }
        return String(data: data, encoding: .utf8) ?? ""
    }

    static func write(_ value: String) {
        let query: [String: Any] = [kSecClass as String: kSecClassGenericPassword, kSecAttrService as String: service, kSecAttrAccount as String: account]
        SecItemDelete(query as CFDictionary)
        guard !value.isEmpty else { return }
        var item = query
        item[kSecValueData as String] = Data(value.utf8)
        item[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        SecItemAdd(item as CFDictionary, nil)
    }
}

File: android/app/src/main/java/com/llmhub/llmhub/agent/McpClient.kt
package com.llmhub.llmhub.agent

import android.content.Context
import com.llmhub.llmhub.BuildConfig
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicLong

data class McpTool(
    val name: String,
    val description: String,
    val inputSchema: JSONObject
) {
    val exposedName: String = "mcp_" + name.replace(Regex("[^A-Za-z0-9_]"), "_")

    fun promptLine(): String {
        val properties = inputSchema.optJSONObject("properties") ?: JSONObject()
        val required = inputSchema.optJSONArray("required")?.let { array ->
            (0 until array.length()).mapNotNull { array.optString(it) }.toSet()
        } ?: emptySet()
        val args = properties.keys().asSequence().map { key ->
            val type = properties.optJSONObject(key)?.optString("type", "value") ?: "value"
            "$key: $type${if (key in required) " (required)" else ""}"
        }.joinToString(", ")
        return "- $exposedName({$args}): ${description.ifBlank { name }.take(300)}"
    }
}

data class McpSettings(
    val enabled: Boolean,
    val transport: String,
    val url: String,
    val token: String,
    val command: String
)

/** Minimal MCP Streamable HTTP client for mobile Agent tool discovery and calls. */
class McpClient(private val context: Context) {
    companion object {
        private const val CURRENT_PROTOCOL = "2026-07-28"
        private const val LEGACY_PROTOCOL = "2025-06-18"
    }

    private val prefs = EncryptedSharedPreferences.create(
        context,
        "agent_mcp_secure",
        MasterKey.Builder(context).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build(),
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
    )
    private val http = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(45, TimeUnit.SECONDS)
        .writeTimeout(15, TimeUnit.SECONDS)
        .build()
    private val ids = AtomicLong(1)
    private var sessionId: String? = null
    private var protocolVersion = CURRENT_PROTOCOL

    var tools: List<McpTool> = emptyList()
        private set

    private fun clientInfo(name: String = "LLM Hub Android") =
        JSONObject().put("name", name).put("version", BuildConfig.VERSION_NAME)

    fun loadSettings() = McpSettings(
        enabled = prefs.getBoolean("enabled", false),
        transport = prefs.getString("transport", "termux") ?: "termux",
        url = prefs.getString("url", "") ?: "",
        token = prefs.getString("token", "") ?: "",
        command = prefs.getString("command", "") ?: ""
    )

    fun saveSettings(settings: McpSettings) {
        prefs.edit().putBoolean("enabled", settings.enabled)
            .putString("transport", settings.transport)
            .putString("url", settings.url.trim())
            .putString("token", settings.token.trim())
            .putString("command", settings.command.trim()).apply()
        if (!settings.enabled) disconnect()
    }

    fun disconnect() {
        sessionId = null
        protocolVersion = CURRENT_PROTOCOL
        tools = emptyList()
    }

    suspend fun connectAndDiscover(settings: McpSettings = loadSettings()): List<McpTool> = withContext(Dispatchers.IO) {
        require(settings.enabled) { "MCP is disabled" }
        if (settings.transport == "termux") {
            require(settings.command.isNotBlank()) { "MCP Termux command is empty" }
            tools = parseTools(termuxRpc(settings.command, "tools/list", JSONObject()))
            return@withContext tools
        }
        validateUrl(settings.url)
        disconnect()

        // Current MCP is stateless. If a server still requires the 2025 lifecycle,
        // initialize it and retain its session ID before listing tools.
        val first = runCatching { rpc(settings, "tools/list", JSONObject(), CURRENT_PROTOCOL) }
        val result = if (first.isSuccess) {
            first.getOrThrow()
        } else {
            protocolVersion = LEGACY_PROTOCOL
            val init = rpc(
                settings,
                "initialize",
                JSONObject()
                    .put("protocolVersion", LEGACY_PROTOCOL)
                    .put("capabilities", JSONObject())
                    .put("clientInfo", clientInfo("LLM Hub")),
                LEGACY_PROTOCOL,
                includeProtocolHeader = false
            )
            protocolVersion = init.optString("protocolVersion", LEGACY_PROTOCOL)
            notify(settings, "notifications/initialized")
            rpc(settings, "tools/list", JSONObject(), protocolVersion)
        }
        tools = parseTools(result)
        tools
    }

    suspend fun callTool(name: String, arguments: JSONObject): String = withContext(Dispatchers.IO) {
        val settings = loadSettings()
        require(settings.enabled) { "MCP is disabled" }
        val result = if (settings.transport == "termux") {
            termuxRpc(settings.command, "tools/call", JSONObject().put("name", name).put("arguments", arguments))
        } else rpc(
            settings,
            "tools/call",
            JSONObject().put("name", name).put("arguments", arguments),
            protocolVersion,
            mcpName = name
        )
        val isError = result.optBoolean("isError", false)
        val output = buildList {
            val content = result.optJSONArray("content") ?: JSONArray()
            for (i in 0 until content.length()) {
                val item = content.optJSONObject(i) ?: continue
                when (item.optString("type")) {
                    "text" -> item.optString("text").takeIf { it.isNotBlank() }?.let(::add)
                    "resource_link" -> item.optString("uri").takeIf { it.isNotBlank() }?.let(::add)
                    else -> add(item.toString())
                }
            }
            result.opt("structuredContent")?.let { add(it.toString()) }
        }.joinToString("\n").ifBlank { result.toString() }
        if (isError) throw McpException(output)
        output.take(100_000)
    }

    /**
     * Termux RUN_COMMAND is one-shot rather than an attachable pipe. Each approved
     * request starts the configured stdio server, performs the MCP lifecycle and
     * one operation, then closes stdin. This keeps stdio support compatible with
     * Android's app sandbox without requiring LLM Hub to execute server binaries.
     */
    private suspend fun termuxRpc(command: String, method: String, params: JSONObject): JSONObject {
        val initId = ids.getAndIncrement()
        val requestId = ids.getAndIncrement()
        val messages = listOf(
            JSONObject().put("jsonrpc", "2.0").put("id", initId).put("method", "initialize").put(
                "params", JSONObject().put("protocolVersion", LEGACY_PROTOCOL).put("capabilities", JSONObject())
                    .put("clientInfo", clientInfo())
            ),
            JSONObject().put("jsonrpc", "2.0").put("method", "notifications/initialized"),
            JSONObject().put("jsonrpc", "2.0").put("id", requestId).put("method", method).put("params", params)
        ).joinToString("\n", postfix = "\n")
        val encoded = android.util.Base64.encodeToString(messages.toByteArray(), android.util.Base64.NO_WRAP)
        // RUN_COMMAND does not initialize Termux's full interactive environment.
        // In particular, npm launchers commonly use `#!/usr/bin/env node`, which
        // needs libtermux-exec to rewrite Linux shebang paths to Termux paths.
        val prefix = "/data/data/com.termux/files/usr"
        val home = "/data/data/com.termux/files/home"
        val shell = "export PREFIX='$prefix' HOME='$home' PATH='$prefix/bin:/system/bin' " +
            "LD_PRELOAD='$prefix/lib/libtermux-exec.so'; " +
            "printf '%s' '$encoded' | base64 -d | timeout 45s $command"
        val output = try {
            TermuxCommandBridge.run(context, shell, 50_000)
        } catch (_: kotlinx.coroutines.TimeoutCancellationException) {
            throw McpException("termux_command_timed_out")
        }
        val response = output.lineSequence().mapNotNull { line ->
            val start = line.indexOf('{')
            if (start < 0) null else runCatching { JSONObject(line.substring(start)) }.getOrNull()
        }.lastOrNull { it.opt("id")?.toString() == requestId.toString() }
            ?: throw McpException("MCP server returned no response: ${output.takeLast(500)}")
        response.optJSONObject("error")?.let { throw McpException(it.optString("message", "MCP request failed")) }
        return response.optJSONObject("result") ?: throw McpException("Invalid MCP response")
    }

    private fun validateUrl(raw: String) {
        val url = raw.trim().toHttpUrlOrNull() ?: throw McpException("Invalid MCP server URL")
        if (url.scheme != "https" && !isPrivateNetworkHost(url.host)) {
            throw McpException("MCP servers must use HTTPS")
        }
    }

    private fun isPrivateNetworkHost(host: String): Boolean {
        val normalized = host.lowercase()
        if (normalized in setOf("localhost", "::1", "10.0.2.2") || normalized.endsWith(".local")) {
            return true
        }
        if (normalized.startsWith("fe80:") || normalized.startsWith("fc") || normalized.startsWith("fd")) {
            return true
        }

        val octets = normalized.split('.').map { it.toIntOrNull() }
        if (octets.size != 4 || octets.any { it == null || it !in 0..255 }) return false
        val first = octets[0]!!
        val second = octets[1]!!
        return first == 10 ||
            first == 127 ||
            (first == 169 && second == 254) ||
            (first == 172 && second in 16..31) ||
            (first == 192 && second == 168)
    }

    private fun parseTools(result: JSONObject): List<McpTool> {
        val array = result.optJSONArray("tools") ?: JSONArray()
        return (0 until array.length()).mapNotNull { index ->
            val item = array.optJSONObject(index) ?: return@mapNotNull null
            val name = item.optString("name").takeIf { it.isNotBlank() } ?: return@mapNotNull null
            McpTool(name, item.optString("description"), item.optJSONObject("inputSchema") ?: JSONObject().put("type", "object"))
        }.take(64)
    }

    private fun rpc(
        settings: McpSettings,
        method: String,
        params: JSONObject,
        version: String,
        includeProtocolHeader: Boolean = true,
        mcpName: String? = null
    ): JSONObject {
        val id = ids.getAndIncrement()
        if (version == CURRENT_PROTOCOL) {
            params.put("_meta", JSONObject().put("io.modelcontextprotocol/clientInfo", clientInfo()))
        }
        val body = JSONObject().put("jsonrpc", "2.0").put("id", id).put("method", method).put("params", params)
        val response = execute(settings, body, method, version, includeProtocolHeader, mcpName)
        response.optJSONObject("error")?.let { throw McpException(it.optString("message", "MCP request failed")) }
        return response.optJSONObject("result") ?: throw McpException("Invalid MCP response")
    }

    private fun notify(settings: McpSettings, method: String) {
        val body = JSONObject().put("jsonrpc", "2.0").put("method", method)
        execute(settings, body, method, protocolVersion, true, null, expectResponse = false)
    }

    private fun execute(
        settings: McpSettings,
        json: JSONObject,
        method: String,
        version: String,
        includeProtocolHeader: Boolean,
        mcpName: String?,
        expectResponse: Boolean = true
    ): JSONObject {
        val builder = Request.Builder().url(settings.url.trim())
            .post(json.toString().toRequestBody("application/json; charset=utf-8".toMediaType()))
            .header("Accept", "application/json, text/event-stream")
            .header("Mcp-Method", method)
        if (includeProtocolHeader) builder.header("MCP-Protocol-Version", version)
        if (!mcpName.isNullOrBlank()) builder.header("Mcp-Name", mcpName)
        sessionId?.let { builder.header("Mcp-Session-Id", it) }
        if (settings.token.isNotBlank()) builder.header("Authorization", "Bearer ${settings.token}")

        http.newCall(builder.build()).execute().use { response ->
            response.header("Mcp-Session-Id")?.let { sessionId = it }
            if (!response.isSuccessful) throw McpException("HTTP ${response.code}: ${response.body?.string()?.take(500).orEmpty()}")
            if (!expectResponse || response.code == 202) return JSONObject()
            val raw = response.body?.string().orEmpty()
            if (raw.length > 1_000_000) throw McpException("MCP response is too large")
            val payload = if (response.header("Content-Type").orEmpty().contains("text/event-stream")) {
                raw.lineSequence().filter { it.startsWith("data:") }.map { it.removePrefix("data:").trim() }
                    .lastOrNull { it.contains("\"jsonrpc\"") } ?: throw McpException("Empty MCP event stream")
            } else raw
            return JSONObject(payload)
        }
    }
}

class McpException(message: String) : Exception(message)
